[Core] Fix #15558: Direct expression in ternary operator

mnesarco opened the pull request:
Changed Expression Grammar to support direct expressions as ternary operator conditions.

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->
Fix #15558

## Previous discussions
https://github.com/FreeCAD/FreeCAD/pull/22506

## Notes
- Backward compatible so no migrations needed for old files. No old models should break.
- Added tests, but more tests are welcome
- Demonstrating Grammar equivalence is a hard problem, tests seem to demonstrate that the new grammar is a superset of the old one.

mnesarco commented:
@hyarion @chennes @pieterhijma @Roy-043 
There is light at the end of the tunnel

hyarion commented:
To my understanding what you do here is:
* Remove `cond` completely
* In ternary op, you use expression instead (`a ? 1mm : 2mm`)
* Allow comparison in any expression (`3mm * (a==2)`)
* Change simplify so that it uses `confusion` instead of `0.5` as threshold value (is this codepath ever used currently?)

It looks like a good change, but I have some questions:

* Looking at `expr` this would allow `STRING` to be used in ternary as well (`"foobar" ? 1mm, 2mm`) and identifiers which is a string.
  - Is this handled currently?
  - Are there any property types that needs to be considered too?
* What happens if you try to use a ternary in another ternary (like `a ? 0 : 1   ?   b ? 2 : 3   :   c ? 4 : 5`)?
* Is it correctly handling precedence correctly with expressions like `a + b ? c - 2 : 3 + d`?

If this has already verified to be working correctly, could you maybe add some tests for it?

mnesarco commented:
> Change simplify so that it uses confusion instead of 0.5 as threshold value (is this codepath ever used currently?)

This path is used, but due to previous grammar structure it worked in a safe way because outputs or `cond` were always 0 or 1. Now that `cond` is gone and `exp` is used, that 0.5 comparison is no longer valid, so `Confusion` is.

pieterhijma commented:
I would also still like to check this out if that's ok.

mnesarco commented on src/App/Expression.y:
```diff
@@ -99,9 +99,15 @@ exp:      num                                   { $$ = $1;
         | exp '%' exp                           { $$ = new OperatorExpression(DocumentObject, $1, OperatorExpression::MOD, $3);   }
         | exp '/' unit_exp                      { $$ = new OperatorExpression(DocumentObject, $1, OperatorExpression::DIV, $3);   }
         | exp '^' exp                           { $$ = new OperatorExpression(DocumentObject, $1, OperatorExpression::POW, $3);   }
+        | exp EQ exp                            { $$ = new OperatorExpression(DocumentObject, $1, OperatorExpression::EQ, $3);    }
```
In C++ and C these operators are left associative, while in Python they have a special treatment as a syntax sugar transforming into a conjunction like `a < b < c` will be rewritten as  `(a < b) and (b < c)`, but this is not just a precedence adjustment but a rewrite rule.

In the old grammar, `= 2<3<4<5` was not supported at all, so at least there is no compatibility issue.

The options is to left them as left associative, or rewrite the expression grammar to reject that kind of expression. I don't think we have to support the python rewrite rule, but the discussion is open.

mnesarco replied:
@hyarion fyi.

mnesarco replied:
As the grammar was updated with @pieterhijma requirements, I will mark this review comment as resolved.

mnesarco commented on src/App/Expression.y:
```diff
@@ -61,16 +61,16 @@ std::stack<FunctionExpression::Function> functions;                /**< Function
-     %precedence EQ NEQ LT GT GTE LTE
-     %precedence ':'
+     %right '?' ':'
+     %left EQ NEQ LT GT GTE LTE
```
See below to keep discussion in one place.

mnesarco commented:
> Regarding chaining, implicit chaining such as a < b < c is now not possible, which is good, but the problem has not been fully solved as explicit chaining, for example (a < b) < c is still possible, whereas this wasn't possible in the previous grammar.

I think that explicit chaining should be allowed as it has clear intention. Specially in cases like `(2 < 3) == (4 < 5)`

mnesarco commented:
I have been doing some tests and experimenting a bit, my conclusion is that there are four options:
1. **The practical one**: revert the last commit and keep the relops left associative like in C/C++
     This will not break any existing model and is consistent with C like languages.
2. **The unicorn**: Structural rewrite of the grammar
    We can get a better grammar but I am not so sure that we can guarantee backwards compatibility and the overall perceived risk is higher. But if the grammar is done right, It will be a future proof and elegant solution.
3. **The lateral**: Resurrect the `if(cond; true_part; false_part)` function proposal
    This is a low risk thing and does not have precedence or associativity problems. We can keep the ternary op unchanged for compatibility but deprecated in docs.
4. **The ugly**: Abandon this idea and live with the existing ternary operator limitations.

pieterhijma commented:
> > For example, I now noticed that FreeCAD's power ^ operator is left associative. I believe that right-associative is typical.
> 
> As I recall it, ^ operator is left associative in spreadsheet applications.

Oh yes, weird...
 
> I agree that it is good to limit the grammar when possible, but what you ask for sounds like quite a big change. Is it worth it or could we just note this as an undefined behavior which might change?

I think that stuff like this is really tricky. I think it is worth to at least try, just to get a better understanding. If we succeed, I would be in favor of the layered grammar. However, as I said, I also can understand the viewpoint that @mnesarco marked as "The practical one". I prefer to stay in the role of reviewer and point out these problems, I let the maintainers decide :)  However, it is good to make an informed decision, in my opinion.

mnesarco commented:
@pieterhijma , @hyarion 
I have added a commit with the grammar rewrite. Test cases are passing and user interaction in Spreadsheet GUI is consistent (Gui behaves different than unit test because the expression is rewritten in gui, so the special case in _toString)

**Summary:**
1. Grammar rewritten in layers
2. _toString fixed to take care of explicit relop chaining for cases like `(a < b) == (c < d)` which is accepted while `a < b == c < d` is not.
3. `(a < b) < c` is still accepted, but I strongly believe that this is ok as it is explicit and it may express a legit intention from the user. `a < b < c` is not accepted as it is implicit and we removed left associativity of relops.

**Tests:**

Passing, but now that the grammar have a significant change, maybe we need more and better tests.

mnesarco commented:
Hello @pieterhijma 

> I do agree that something like `(a < b) == (c < d)` can express a legitimate intent and that it would be good to support it. So, this means that I switch my standpoint from being conservative (generalize while keeping the restrictions as much as possible) to accepting generalizing conditions to expressions in full. However, in that case, I would prefer option 1 that simply allows implicit chaining as well, keeping the changes to the grammar minimal and let the precedence decide the semantics of chaining, similar to in C.

Thank you for reviewing again. The good thing is that your independent conclusions brings confidence to the path forward.
 
> In practice, this boils down to reverting last two commits. It would be nice though, to have test cases that show that we allow these cases.
 
I will add a revert commit with the additional test cases. At the end the PR must be squashed but until that I prefer to add  commits following the process.

> Sorry for backtracking on this, but I believe that these different implementations allowed me (and perhaps you and @hyarion as well) to understand the requirements better.

This kind of problems are a bit abstract without an actual grammar to test, so the backtracking is part of the process.

mnesarco commented:
> Hello! I don't know where to ask this question. Is there a way to create multiple conditions without nesting? For example, I'm checking for equalities like A=1 B=2 C=3 D=4, and so on. I have to write this: A==1 ? 1 : (B==2 ? (C==3 ? 3:4))
> 
> Maybe there's a more convenient way to check a chain of values ​​(or is there always only two?). Sorry if this is a stupid question—I'm not a programmer. I thought the ternary operator was what I needed, but I couldn't find any examples with more than two values.
> 
> It is very difficult to form such expressions with property names. Type_ID_Info == 1 ? Z_PARTS_id1 : (Type_ID_Info == 2 ? Z_PARTS_id2 : Z_PARTS_id3)

It is not supported right now by FreeCAD expressions.

I think it can be added in the same way as excel manages that case, two functions came to mi mind:

Choose function:
https://support.microsoft.com/en-us/office/choose-function-fc5c184f-cb62-4ec7-a46e-38653b98f5bc

Switch function:
https://support.microsoft.com/en-au/office/switch-function-47ab33c0-28ce-4530-8a45-d532ec4aa25e

Some ideas:
```
=choose(expr1, result1, expr2, result2, expr3, result3, ..., default)
=match(expr, val1, result1, val2, result2, ..., default)
=if(cond1, result1, else1, resrult2, else2, result3, ..., default)
```

@pieterhijma @hyarion what do you think?

hyarion commented:
> @pieterhijma @hyarion what do you think?

I would recommend creating an issue for it and continue the discussion there instead
